### DISCUSSION TITLE
Why do I get a message that all messages must have non-empty content?

### DISCUSSION THREAD

**Discussion by madorian** [karthink/gptel#406] [General]
I get this message, like 10 times per day. 

```
Claude error: (HTTP/2 400) messages.8: all messages must have non-empty content except for the optional final assistant message
```

I don't really do anything weird;). I've tried to search and it seems there are lots of people who get this, but I haven't found it in relation to gptel. Is there anything I can do to avoid this?

**Comment by karthink** [comment]
Yeah, this is a problem with gptel that I don't know how to fix yet.  You are probably modifying the responses from the model in some way, by doing something like filling the region, and that is causing the response to be split into a few separate responses separated by empty user prompts.

Multiple fixes have been discussed (see #321) but all of them cause something else to break.  If you're able to, try switching to the `feature-overlays` branch of gptel, which doesn't have this problem (but has other problems).

**Comment by karthink** [comment]
I've pushed a fix, please try now.

**Comment by madorian** [comment]
I've installed like this: 
```
(straight-use-package 'gptel)
```
Is there some way to check what revision is installed?

**Reply by karthink** [reply]
The easiest thing to do would be to run `M-x straight-update-package` `gptel`.  That will ensure you're on the latest commit.

To look up the installed version: I don't think straight tells you that.  You can open up `gptel.el` (`M-x find-library` `gptel`) and look at the header, if it says version 0.9.6 you're on the latest.

**Comment by madorian** [comment]
I don't have a straight-update-package command. I did find straight-pull-package, so now I'm on 0.9.6;). Thanks. I'll report back

**Comment by karthink** [comment]
> I don't have a straight-update-package command. I did find straight-pull-package

That's the one, sorry.

**Comment by madorian** [comment]
I have not seen this message after this fix;). Thanks a bunch.

**Comment by madorian** [comment]
Hmm, I haven't seen it since, but today I got it again;)

Claude error: (HTTP/2 400) messages.6: all messages must have non-empty content except for the optional final assistant message

If it's just one in a million, then I guess it's not biggie, but just to let you know;)

Also, what is the proper way to reset? I mean, what I do now, is that I have to exit Emacs to be able to use gptel again.

**Comment by madorian** [comment]
Hmm, got a new one today. 

Claude error: (HTTP/2 400) messages: final assistant content cannot end with trailing whitespace

**Comment by karthink** [comment]
> Hmm, got a new one today. 
>
> Claude error: (HTTP/2 400) messages: final assistant content cannot end with trailing whitespace

Thanks, looking into it.

**Reply by madorian** [reply]
Yeah, I believe I'm on HEAD, but is there a gptel-version or something?

This happens many times a day, now

When it happens, is there any way to reset it, instead of closing down Emacs?;)

**Reply by karthink** [reply]
This issue appears to be fixed for everyone else who was experiencing it.  A reproducible test case can help me identify the problem.  Could you do the following?

1. Make sure that `gptel-mode` is on.  If this is a chat buffer it's on by default.
2. When you get this error, save the chat to a file.  (markdown or org-mode)
3. Share the file here.

**Comment by madorian** [comment]
[chat.md](https://github.com/user-attachments/files/18160833/chat.md)

Is there supposed to be any debug info in this file?

**Comment by madorian** [comment]
Still getting that message now and then and right now;)

I assume this is the same issue. 

```
Claude error: ((HTTP/2 400) invalid_request_error) messages: final assistant content cannot end with trailing whitespace
```

I use ```(straight-use-package 'gptel)```, but is there maybe an M-x gptel-version or something to verify?

I also cannot reset it, by closing the buffer, so that means I cannot start any new chats. I have to restart the whole of Emacs, so any function to reset it would be great;)

**Comment by karthink** [comment]
> Still getting that message now and then and right now;)

Please provide a reproducible recipe.

**Comment by madorian** [comment]
It has happened five times today, but I cannot really give a reproducible recipe;)

I think, first of all, I need to know what that error message is saying, cause I don't understand what the meaning is. 

It says it cannot end with a trailing whitespace; what does that even mean? It cannot have whitespace after the last character?

**Comment by karthink** [comment]
> I think, first of all, I need to know what that error message is saying, cause I don't understand what the meaning is. 
>
> It says it cannot end with a trailing whitespace; what does that even mean? It cannot have whitespace after the last character?

I think it means that there is a user message that is only whitespace.  This may or may not be the last message.

If you can save a chat where this happens to a file and include the file here I can take a look.

**Comment by madorian** [comment]
Here's a chat where I got this message, so now I need to restart Emacs.

A very small chat. It usually doesn't happen this fast. 

```
###             var lookup = flatList.ToDictionary(d => d.Id.Value);

warning CS8629: Nullable value type may be null. 


This warning is indicating a potential null reference issue in your C# code. The specific problem is:

1. You're trying to access `d.Id.Value` in the lambda expression.
2. `Id` is likely a nullable type (probably `int?` or similar).
3. The compiler is warning that `Id` might be null when you try to access its `Value` property.

To fix this, you have a few options:

1. Use null-conditional operator:
   -```csharp
   var lookup = flatList.ToDictionary(d => d.Id?.Value);
   -```

2. Filter out null values first:
   -```csharp
   var lookup = flatList.Where(d => d.Id.HasValue).ToDictionary(d => d.Id.Value);
   -```

3. If you're certain `Id` will never be null, you can use the null-forgiving operator:
   -```csharp
   var lookup = flatList.ToDictionary(d => d.Id!.Value);
   -```

Choose the option that best fits your specific use case and data guarantees.

### 

<!-- Local Variables: -->
<!-- gptel-model: claude-3-5-sonnet-20240620 -->
<!-- gptel--backend-name: "Claude" -->
<!-- gptel--bounds: ((185 . 1042) (1241 . 1298)) -->
<!-- End: -->

explain:
   var lookup = flatList.ToDictionary(d => d.Id?.Value);

```

EDIT: I had to put a '-' in front of lines starting with three "'"

**Reply by karthink** [reply]
Why do you have text below the local variables block?

**Reply by madorian** [reply]
It's probably because I use
```
C-<end> runs the command end-of-buffer (found in global-map)
```

, so I end up below this block. 

I'm sure there is a command to move to the last prompt?

Does this block need to be at the bottom? We can't put it at the top or are we supposed to be able to change that in a chat?

**Reply by karthink** [reply]
> Does this block need to be at the bottom? We can't put it at the top or are we supposed to be able to change that in a chat?

Yes, it must.  See `(info "(emacs) Specifying File Variables")`

> I'm sure there is a command to move to the last prompt?

There's `gptel-end-of-response` to go to the end of the current response.  You can give it a prefix argument to repeat.

I would stop clobbering the local variables list first.  After that if the "non-empty content" bug reappears I can look at it.

**Reply by madorian** [reply]
ok, it didn't help that I always write on the prompt. 

```
### #+begin_src elisp
(defun copy-region-as-append ()
  "Copy the selected region to the kill ring, appending it to the last kill."
  (interactive)
  (if (use-region-p)
      (let ((begin (region-beginning))
            (end (region-end)))
        (if (eq last-command 'kill-region)
            (kill-append (buffer-substring begin end) (< end begin))
          (kill-new (buffer-substring begin end)))
        (setq deactivate-mark t)
        (message "Region appended to kill ring"))
    (message "No region active")))

;; Bind the function to a key combination, e.g., C-c C-a
(global-set-key (kbd "C-c C-a") 'copy-region-as-append)
#+end_src


1. It checks if there's an active region using =use-region-p=.
2. If there's an active region:
   - It gets the beginning and end of the region.
   - If the last command was =kill-region=, it appends the new text to the last kill using =kill-append=.
   - Otherwise, it creates a new kill ring entry using =kill-new=.
   - It deactivates the mark (unselects the region).
   - It displays a message confirming the action.
3. If there's no active region, it displays a message saying so.

I don't understand the following: 

If the last command was =kill-region=, it appends the new text to the last kill using =kill-append=.
   - Otherwise, it creates a new kill ring entry using =kill-new=.


<!-- Local Variables: -->
<!-- gptel-model: claude-3-5-sonnet-20240620 -->
<!-- gptel--backend-name: "Claude" -->
<!-- gptel--bounds: ((5 . 1169) (1171 . 1339)) -->
<!-- End: -->

```

**Reply by madorian** [reply]
Hmm, I don't actually see a prompt in there

**Reply by madorian** [reply]
Ah, ok, cause I hadn't even sent it
So, this actually triggered the bug, even before I had gotten a first reply;)

**Comment by gcv** [comment]
For what it's worth, I'm getting a similar error from gptel: `Claude error: (HTTP/2 400) messages: text content blocks must contain non-whitespace text`.

All I did was put a chunk of Python code in the prompt. It looks like this:

    >>> Review this code:
    ```
    def my_fun():
       ...
    ```

It's not reproducible, except in a chat where it actually happens. Once it happens, pasting code into the prompt pretty much stops working, but non-code prompts continue to work. I'm sorry, I cannot post this exact chat.

**Reply by madorian** [reply]
When this happens to you, do you have to restart Emacs to continue working?

**Reply by gcv** [reply]
Ooof, I don't remember, sorry! It's been a while since I used gptel for programming directly. I switched to Claude Code for daily work (the convenience of it managing context has been a huge time saver). I still use gptel for its chat buffer. It's been on my list of things to try gptel again more for programming and use its buffer context features.

I don't _think_ I had to restart Emacs, but I definitely had to kill the gptel chat buffer and lose context.
